app/gui2/public/visualizations/measurement.ts

```diff
@@ -0,0 +1,19 @@
+/** Helper function to get text width to make sure that labels on x axis do not overlap,
+ * and keeps it readable. */
+export function getTextWidth(
+  text: string | null | undefined,
+  fontSize = '11.5px',
+  fontFamily = "Inter, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Fira Sans', 'Droid Sans', 'Helvetica Neue', sans-serif",
+) {
+  if (text == null) {
+    return 0
+  }
+  const canvas = document.createElement('canvas')
+  const context = canvas.getContext('2d')
+  if (context == null) {
+    throw new Error('Could not get canvas 2D context.')
+  }
+  context.font = `${fontSize} ${fontFamily}`
```

Comment:
You don't need to create a new canvas each time. Doing it once lazily should be better. Once you have a context, you can reuse it for multiple text measurements.

```ts
let _measureContext: CanvasRenderingContext2D = null;
function getMeasureContext() {
  if (_measureContext == null) {
    const canvas = document.createElement('canvas')
    _measureContext = canvas.getContext('2d') ?? bail('Could not get canvas 2D context.')
  }
  return _measureContext 
}

// in getTextWidth
const context = getMeasureContext()
```